File: src-ui/lib/log.ts
import { bind } from "./bindings";

export function debug<T>(content: T): T {
  console.debug(content);
  return content;
}

export function info(...message: any[]) {
  console.info(...message);
}

export function warn(...message: any[]) {
  console.warn(...message);
}

export function error(...message: any[]) {
  console.error(...message);
}

bind(function logInfo(message: string) {
  info(message);
});
bind(function logWarn(message: string) {
  warn(message);
});
bind(function logError(message: string) {
  error(message);
});

File: src-tauri/src/wasm/log.rs
use crate::{js::call_js_or_default, wasm};

wasmtime::component::bindgen!({
  world: "logger",
  imports: { default: async },
  exports: { default: async},
});

impl pato::plugin::log::Host for wasm::HostData {
    async fn info(&mut self, message: String) {
        call_js_or_default::<_, ()>("logInfo", message).await
    }

    async fn warn(&mut self, message: String) {
        call_js_or_default::<_, ()>("logWarn", message).await
    }

    async fn error(&mut self, message: String) {
        call_js_or_default::<_, ()>("logError", message).await
    }
}
